### ISSUE TITLE
Exclude/Includes breaks recursive folder reading

### ISSUE THREAD

**Issue by balinju** [joelittlejohn/jsonschema2pojo#332]
I have the following configuration:

```
<groupId>org.jsonschema2pojo</groupId>
<artifactId>jsonschema2pojo-maven-plugin</artifactId>
<version>0.4.10</version>
<configuration>
	<sourceDirectory>${basedir}/src/main/api/schemas</sourceDirectory>
	<targetPackage>com.xxx.domain</targetPackage>
	<outputDirectory>${project.build.directory}/generated/jsonschema</outputDirectory>
	<removeOldOutput>true</removeOldOutput>
	<annotationStyle>jackson1</annotationStyle>
</configuration>
```

This breaks on Linux due to the .svn folders

```
[ERROR] Failed to execute goal org.jsonschema2pojo:jsonschema2pojo-maven-plugin:0.4.10:generate (default) on project p1: Execution default of goal org.jsonschema2pojo:jsonschema2pojo-maven-plugin:0.4.10:generate failed: Error parsing document: file:/opt2/jenkins-data/jobs/Build-p1/workspace/src/main/api/schemas/.svn/all-wcprops: Unexpected character ('K' (code 75)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')
```

Temporary removing .svn folder to test works brilliantly.

Adding Exclude/Include to only process *.json as shown below uncovers another issue

```
<includes>
	<include>*.json</include> 
</includes>
```

When the above is added, only the schemas at the top level of the directory are read. The directory is not recursively read. This was working without the include. I also tried the below, but did not work either.

```
<includes>
	<include>**/*.json</include> 
</includes>
```

**Comment by balinju** [comment]
Hi,
I have checked this issue out and I have discovered the source of the problem. When no includes/excludes are configured, the AllFileFilter is used which returns true on every check. However when include/excludes are configured, the MatchPatternsFileFilter is used instead. The problem here is with the directory check against include/excludes.

In the accept() method, !excludePatterns.matchesPatternStart(path, caseSensitive) is always returning false if the exclude list contains at least one item. By default the createFilter() method adds the default exclude list so this list is never empty.

This might be a misuse of the method matchesPatternStart().

**Comment by joelittlejohn** [comment]
This is an interesting case, thanks for digging in to the code! One option would be to simply apply includes/excludes to files and not directories. Some people might be dedending on the directory check though. Another, more powerful but also more complex option would be to support a proper directory match syntax (I would have expected the same ** syntax you have given above).

@ctrimble might have an opinion on the best course of action here.

**Comment by balinju** [comment]
That's what I did at the moment for a workaround, skipping directory check and base the checks at the file level. However I do agree with you that proper directory matching syntax would be the best solution.

**Comment by ctrimble** [comment]
@balinju I am sorry that you ran into issues with the file matching code.  It seems like we have 2 distinct issues here.  

__Issue 1:__ With no includes/excludes defined, the matching should be as though this was configured:

```
<includes>
  <include>**/*</include>
</includes> 
```

__Issue 2:__ Patterns containing `**` are not recursing the file system.

Do these two items properly sum up the issues you are seeing?

**Comment by ctrimble** [comment]
One other note, the pattern `*.json` should only match items in the top level directory.  You need the `**` pattern to match files nested in directories.

**Comment by ctrimble** [comment]
#340 should take care of the send issue.

@joelittlejohn should I submit a PR to exclude directories like `.svn` by default?  It seems like the behavior that every reasonable user would expect and want, but would narrow the possible directory structures.

**Comment by joelittlejohn** [comment]
Agreed. I think there's an argument for ignoring .* by default, but that
might be over-reaching. Ignoring .svn by default is a good feature IMO.

Thanks a lot for submitting these changes btw! Excellent stuff.

**Comment by balinju** [comment]
Yeah agreed too. Thanks for the fix.

**Comment by ctrimble** [comment]
@joelittlejohn and @balinju I am going to get this other change into #340 and squash it into 2 commits, one for each issue.  I will comment there when it is ready to merge.

**Comment by ctrimble** [comment]
@balinju thanks for reporting this.  Your diagnosis was spot on.  I was filtering things that could possibly match an exclude pattern, instead of things that definitely matched an exclude pattern.

**Comment by balinju** [comment]
@ctrimble no problem :) Happy to have contributed.

**Comment by joelittlejohn** [comment]
Closed by #340.
